lib_gb/src/mmu/vram_dma_controller.rs

```diff
@@ -0,0 +1,132 @@
+use crate::{utils::bit_masks::BIT_7_MASK, ppu::{gb_ppu::GbPpu, gfx_device::GfxDevice, ppu_state::PpuState}};
+
+use super::external_memory_bus::ExternalMemoryBus;
+
+enum TransferMode{
+    GeneralPurpose,
+    Hblank,
+    Terminated
+}
+
+const HBLANK_TRANSFER_CHUNK_SIZE:u8 = 0x10;
+const GP_TRASNFER_BYTES_PER_MCYCLE:u8 = 2;
+
+pub struct VramDmaController{
+    source_address:u16,
+    dest_address:u16,
+    mode:TransferMode,
+    remaining_length:u8,
+
+    last_ly:u8,
+    m_cycle_counter:u32
+}
+
+impl VramDmaController{
+    pub fn new()->Self{
+        Self{
+            dest_address: 0,
+            last_ly: 0,
+            m_cycle_counter: 0,
+            mode: TransferMode::Terminated,
+            remaining_length: 0,
+            source_address: 0
+        }
+    }
+
+    pub fn set_source_high(&mut self, value:u8){
+        self.source_address = (self.source_address & 0x00FF) | (value as u16) << 8;
+    }
+    
+    pub fn set_source_low(&mut self, value:u8){
+        // Ignores the last 4 bits of the source
+        let value = value & 0xF0;
+        self.source_address = (self.source_address & 0xFF00) | value as u16;
+    }
+
+    pub fn set_dest_high(&mut self, value:u8){
+        // Upper 3 bits are ignored since the dest are always in the vram
+        let value = value & 0b0001_1111;
+        self.dest_address = (self.dest_address & 0x00FF) | (value as u16) << 8;
+    }
+    
+    pub fn set_dest_low(&mut self, value:u8){
+        // Ignores the last 4 bits of the dest
+        let value = value & 0xF0;
+        self.dest_address = (self.dest_address & 0xFF00) | value as u16;
+    }
+
+    pub fn set_mode_length(&mut self, value:u8){
+        match self.mode{
+            TransferMode::Hblank |
+            TransferMode::GeneralPurpose=>self.mode = TransferMode::Terminated,
```

Comment:
This is not correct,

the GP should not be reachable (maybe panic?) and HBlank should terminate only on bit 7 == 0